build/purger/purger.py

```diff
@@ -0,0 +1,235 @@
+)
+SYNAPSE_CONFIG_PATH = "/config/synapse.yaml"
+
+
+@click.command()
+@click.argument("db_uri", envvar="MATRIX_DB_URI")
+@click.argument("server", envvar="MATRIX_SERVER")
+@click.option("-c", "--credentials-file", required=True, type=click.File("rt"))
+@click.option("--keep-newer", type=int, help="Keep events newer than this number of days")
+@click.option(
+    "--keep-min-msgs",
+    type=int,
+    help="Keep at least this number of message events per room, regardless of --keep-newer",
+)
+@click.option(
+    "--parallel-purges",
+    default=10,
+    show_default=True,
+    type=int,
+    help="Max number of purges to run in parallel.",
+)
+@click.option(
+    "--post-sql",
+    type=click.File("rt"),
+    help=(
+        "Pass a SQL script file as parameter to run it on DB after purging. "
+        'Useful to run cleanup scripts, like "synapse_janitor.sql".'
+    ),
+)
+@click.option(
+    "--docker-restart-label",
+    help="If set, search all containers with given label and, if they're running, restart them",
+)
+def purge(
+    db_uri: str,
+    server: str,
+    credentials_file: TextIO,
+    keep_newer: int,
+    keep_min_msgs: int,
+    parallel_purges: int,
+    post_sql: TextIO,
+    docker_restart_label: str,
+) -> None:
+    """ Purge historic data from rooms in a synapse server
+
+    DB_URI: DB connection string: postgres://user:password@netloc:port/dbname
+    SERVER: matrix synapse server url, e.g.: http://hostname
+
+    All option can be passed through uppercase environment variables prefixed with 'MATRIX_'
+    e.g.: export MATRIX_KEEP_MIN_MSGS=100
+    """
+    session = requests.Session()
+
+    try:
+        credentials = json.loads(credentials_file.read())
+        username = credentials["username"]
+        password = credentials["password"]
+    except (JSONDecodeError, UnicodeDecodeError, OSError, KeyError) as ex:
+        click.secho(f"Invalid credentials file: {ex}", fg="red")
+        sys.exit(1)
+
+    api = GMatrixHttpApi(server)
+    try:
+        response = api.login("m.login.password", user=username, password=password)
+        admin_access_token = response["access_token"]
+    except (MatrixError, KeyError) as ex:
+        click.secho(f"Could not log in to server {server}: {ex}")
+        sys.exit(1)
+
+    try:
+        with psycopg2.connect(db_uri) as db, db.cursor() as cur:
+            purges: Dict[str, str] = dict()
+
+            def wait_and_purge_room(room_id: str = None, event_id: str = None) -> None:
+                """ Wait for available slots in parallel_purges and purge room
+
+                If room_id is None, just wait for current purges to complete and return
+                If event_id is None, purge all events in room
+                """
+                while len(purges) >= (parallel_purges if room_id else 1):
+                    # wait and clear completed purges
+                    time.sleep(1)
```

Comment:
Why is the sleep here necessary? The request below is blocking, right?

Reply:
I'm not sure. 
@andrevmatos originally wrote it. I just moved it to the `build` directory and switched it to use the new server local admin user feature.